ta/pkcs11/src/pkcs11_token.c

```diff
@@ -80,3 +81,31 @@ void pkcs11_deinit(void)
 	for (id = 0; id < TOKEN_COUNT; id++)
 		close_persistent_db(get_token(id));
 }
+
+uint32_t entry_ck_slot_list(uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *out = &params[2];
+	const size_t out_size = sizeof(uint32_t) * TOKEN_COUNT;
+	uint32_t *id = NULL;
+	uint32_t n = 0;
+
+	if (ptypes != exp_pt ||
+	    params[0].memref.size != TEE_PARAM0_SIZE_MIN)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	if (out->memref.size < out_size) {
+		out->memref.size = out_size;
+		return PKCS11_CKR_BUFFER_TOO_SMALL;
+	}
+
+	for (id = out->memref.buffer, n = 0; n < TOKEN_COUNT; n++, id++)
+		TEE_MemMove(id, &n, sizeof(n));
```

Comment:
Using a possibly unaligned pointer like this is breaks with the C standard.
It may work in practise depending on compiler and compiler options, but I've seen breakage in the past.

Reply:
You mean compiler could not consider pointer alignments when copying data.

In former changes for the TA there will be lot a de-serialization of 16bit or 32bit data from client buffers, likely not aligned in the buffer. I expected to use such `memcpy()` like function to do so. It sounds strange to me that such sequence is prone to issues.

Reply:
It's the `uint32_t *id` pointer which is a problem, from the C99 standard 6.3.2.3 Pointer paragram 7:
> A pointer to an object or incomplete type may be converted to a pointer to a different
object or incomplete type. If the resulting pointer is not correctly aligned for the
pointed-to type, the behavior is undefined. Otherwise, when converted back again, the
result shall compare equal to the original pointer. When a pointer to an object is
converted to a pointer to a character type, the result points to the lowest addressed byte of
the object. Successive increments of the result, up to the size of the object, yield pointers
to the remaining bytes of the object.

We can ignore this and it will probably never be a problem, or we could write compliant code by using a `uint8_t *id` instead which is incremented by `sizeof(uint32_t)`.

Reply:
Oh yes, sorry, I misunderstood your comment.
Right, I'll change to `uint8_t *`.